File: tests/r_parity/18_augsynth.py
"""StatsPAI Augmented SCM parity (Python side) -- Module 18.

Runs the **native Python** augmented SCM (``sp.augsynth(backend='native')``)
on the Basque-Country replica (Ben-Michael, Feller & Rothstein 2021).
The companion 18_augsynth.R uses ``augsynth::augsynth`` on the same CSV.

Tier: T2 iterative parity.  The native Python path ports the same
control-mean centering, time-holdout ridge-lambda CV, SCM weights, and
ridge-augmented weight formula used by ``augsynth::augsynth`` for this
no-covariate Ridge+SCM specification.  The optional
``backend='augsynth'`` R bridge remains a migration convenience, NOT the
parity comparator.
"""
from __future__ import annotations

import statspai as sp

from _common import ParityRecord, dump_csv, write_results


MODULE = "18_augsynth"


def main() -> None:
    df = sp.datasets.basque_terrorism()
    dump_csv(df, MODULE)

    fit = sp.augsynth(
        df,
        outcome="gdppc",
        unit="region",
        time="year",
        treated_unit="Basque Country",
        treatment_time=1970,
        backend="native",
    )

    rows: list[ParityRecord] = [
        ParityRecord(
            module=MODULE, side="py", statistic="att_augmented",
            estimate=float(fit.estimate),
            se=float(fit.se),
            n=int(len(df)),
        ),
    ]
    _pre = fit.model_info.get("pre_treatment_rmse",
                              fit.model_info.get("pre_rmspe"))
    if _pre is not None:
        rows.append(
            ParityRecord(
                module=MODULE, side="py", statistic="pre_rmspe",
                estimate=float(_pre),
                n=int(len(df)),
            )
        )

    write_results(
        MODULE, "py", rows,
        extra={
            "method": "augmented",
            "backend": fit.model_info.get("backend", "native"),
            "n_donors": int(fit.model_info.get("n_donors", 0)),
            "tier": "T2",
            "reference_backend": "augsynth",
            "native_parity_note": (
                "Headline row is the NATIVE Python augmented SCM "
                "(backend='native'). It ports augsynth's centered pre-outcome "
                "SCM plus ridge-augmented weight formula and matches "
                "augsynth::augsynth on the Basque fixture within iterative "
                "solver tolerance. The optional backend='augsynth' R bridge "
                "is a migration convenience, not the parity comparator."
            ),
        },
    )


if __name__ == "__main__":
    main()

File: tests/r_parity/18_augsynth.R
# StatsPAI Augmented SCM parity (R side) -- Module 18.
#
# Reads data/18_augsynth.csv and runs augsynth::augsynth.
# Tolerance: rel < 1e-4 on the post-treatment ATT.  The native Python
# row ports augsynth's centered pre-outcome Ridge+SCM weight convention;
# remaining differences are iterative solver tolerance, not a T4 gap.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(augsynth)
})

MODULE <- "18_augsynth"

df <- read_csv_strict(MODULE)
df$treated_indicator <- as.integer(
  df$region == "Basque Country" & df$year >= 1970
)

# augsynth::augsynth API: outcome ~ treatment_indicator | aux_covariates
fit <- augsynth::augsynth(
  form = gdppc ~ treated_indicator,
  unit = region,
  time = year,
  data = df,
  progfunc = "Ridge",
  scm = TRUE
)

# Aggregate to the post-treatment ATT (average of effects).
sm <- summary(fit)
att_est <- sm$average_att$Estimate

# Pre-period RMSPE: compute from the synthetic counterfactual via
# predict(). predict(fit) returns the synthetic-control trajectory
# named by year. Compare against the treated unit's actual outcome.
synth_traj <- predict(fit)
yrs <- as.integer(names(synth_traj))
treated_y <- df$gdppc[df$region == "Basque Country"]
treated_yrs <- df$year[df$region == "Basque Country"]
treated_y_aligned <- treated_y[match(yrs, treated_yrs)]
pre_idx <- yrs < 1970
pre_residuals <- treated_y_aligned[pre_idx] - synth_traj[pre_idx]
pre_rmspe <- sqrt(mean(pre_residuals^2))

rows <- list(
  parity_row(MODULE, "att_augmented",
             estimate = att_est, n = nrow(df)),
  parity_row(MODULE, "pre_rmspe", estimate = pre_rmspe, n = nrow(df))
)

write_results(MODULE, rows,
              extra = list(method = "augsynth::augsynth",
                           progfunc = "Ridge"))
